src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -0,0 +1,227 @@
+        if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(), "Token must be START_OBJECT");
+        }
+        parser.nextToken();
+        neuralQueryBuilder.fieldName(parser.currentName());
+        parser.nextToken();
+        parseQueryParams(parser, neuralQueryBuilder);
+        parser.nextToken();
+        if (parser.currentToken() != XContentParser.Token.END_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(), "Token must be END_OBJECT");
+        }
+
+        requireValue(neuralQueryBuilder.getQueryText(), "Query text must be provided for neural query");
+        requireValue(neuralQueryBuilder.getFieldName(), "Field name must be provided for neural query");
+        requireValue(neuralQueryBuilder.getModelId(), "Model ID must be provided for neural query");
+        requireValue(neuralQueryBuilder.getK(), "K must be provided for neural query");
+
+        return neuralQueryBuilder;
+    }
+
+    private static void parseQueryParams(XContentParser parser, NeuralQueryBuilder neuralQueryBuilder) throws IOException {
+        XContentParser.Token token;
+        String currentFieldName = "";
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (QUERY_TEXT_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    neuralQueryBuilder.queryText(parser.text());
+                } else if (MODEL_ID_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    neuralQueryBuilder.modelId(parser.text());
+                } else if (K_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    neuralQueryBuilder.k((Integer) NumberFieldMapper.NumberType.INTEGER.parse(parser.objectBytes(), false));
+                } else if (NAME_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    neuralQueryBuilder.queryName(parser.text());
+                } else if (BOOST_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    neuralQueryBuilder.boost(parser.floatValue());
+                } else {
+                    throw new ParsingException(
+                        parser.getTokenLocation(),
+                        "[" + NAME + "] query does not support [" + currentFieldName + "]"
+                    );
+                }
+            } else {
+                throw new ParsingException(
+                    parser.getTokenLocation(),
+                    "[" + NAME + "] unknown token [" + token + "] after [" + currentFieldName + "]"
+                );
+            }
+        }
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext queryShardContext) throws IOException {
```

Comment:
"toQuery" will build a Query object using the queryShardContext and the queryBuilder. "doToQuery" is used internally to produce the query in [AbstractQueryBuilder](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractQueryBuilder.java#L116).

Its null right now because we will be implementing in future PR. Ill add a comment.